src/validation.cpp

```diff
@@ -5036,24 +5036,6 @@ CBlockFileInfo* GetBlockFileInfo(size_t n)
     return &vinfoBlockFile.at(n);
 }
 
-ThresholdState VersionBitsTipState(const Consensus::Params& params, Consensus::DeploymentPos pos)
-{
-    LOCK(cs_main);
```

Comment:
In commit "validation: Remove global ::VersionBitsTip{State,SinceHeight,Statistics}" (b60c7e5a5a60fe60a3bc8ba540dc7c8d5c935869)

I'm a little unclear why it's safe to drop these locks, or why they were here to begin with, or if VersionBits or ChainActive or Tip functions should have been annotated to require cs_main. It seems like something should have be annotated if cs_main was required here

Reply:
Digging in the git history a bit, it seems that the earliest of this trio -- namely `VersionBitsTipState` -- was introduced in d23f6c6a0d2. My hypothesis is that the `LOCK(cs_main)` was added in order to stop the chain from advancing and make sure that the caller got info on the actual tip. However, it seems like a `AssertLockHeld(cs_main);` annotation would have been much more suitable in that case (and existed in the codebase at d23f6c6a0d2).

I'm not entirely sure about the original intent here, so perhaps @sipa can enlighten me and make sure I'm not missing something!

Reply:
At least add a comment about the fact the `cs_main` lock is already taken in this code path L1317 (`src/rpc/blockchain.cpp`) ?

Reply:
@ariard @ryanofsky would it be preferable to add a `EXCLUSIVE_LOCKS_REQUIRED(cs_main)` annotation to `VersionBitsTip{State,SinceHeight,Statistics}` here just in case they get new callers in the future?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/20158#discussion_r546119285

In commit "validation: Remove global ::VersionBitsTip{State,SinceHeight,Statistics}" (6ed743d86fb1b5466076eaacf1e8ebc830abde62)

> @ariard @ryanofsky would it be preferable to add a `EXCLUSIVE_LOCKS_REQUIRED(cs_main)` annotation to `VersionBitsTip{State,SinceHeight,Statistics}` here just in case they get new callers in the future?

I'd keep what you have now unless there's a reason to think conclusion from digging into history is incorrect.

It would be confusing to add a lock required annotation if we don't think a lock is required. In theory you could write `EXCLUSIVE_LOCKS_REQUIRED(cs_main)` with a comment "There is no known reason to think a lock is required to call this function, but the original code had an explicit recursive lock, so locking is stilll required to be conservative in case there is an unknown reason locking is needed." This would probably be overkill though.